tests/conftest.py

```diff
@@ -0,0 +1,104 @@
+        self.task_timeout = 60
+        self.task_poll = 1
+        return super(ForemanApi, self).__init__(**kwargs)
+
+    def set_changed(self):
+        pass
+
+    def fail_from_exception(self, exc, msg):
+        raise exc
+
+    def _resource(self, resource):
+        if resource not in self.resources:
+            raise Exception("The server doesn't know about {0}, is the right plugin installed?".format(resource))
+        return self.resource(resource)
+
+    def _resource_call(self, resource, *args, **kwargs):
+        return self._resource(resource).call(*args, **kwargs)
+
+    def _resource_prepare_params(self, resource, action, params):
+        api_action = self._resource(resource).action(action)
+        return api_action.prepare_params(params)
+
+    def resource_action(self, resource, action, params, options=None, data=None, files=None,
+                        ignore_check_mode=False, record_change=True, ignore_task_errors=False):
+        resource_payload = self._resource_prepare_params(resource, action, params)
+        if options is None:
+            options = {}
+        try:
+            result = None
+            if ignore_check_mode or not self.check_mode:
+                result = self._resource_call(resource, action, resource_payload, options=options, data=data, files=files)
+                is_foreman_task = isinstance(result, dict) and 'action' in result and 'state' in result and 'started_at' in result
+                if is_foreman_task:
+                    result = self.wait_for_task(result, ignore_errors=ignore_task_errors)
+        except Exception as e:
+            msg = 'Error while performing {0} on {1}: {2}'.format(
+                action, resource, to_native(e))
+            self.fail_from_exception(e, msg)
+        if record_change and not ignore_check_mode:
+            # If we were supposed to ignore check_mode we can assume this action was not a changing one.
+            self.set_changed()
+        return result
+
+    def wait_for_task(self, task, ignore_errors=False):
+        duration = self.task_timeout
+        while task['state'] not in ['paused', 'stopped']:
+            duration -= self.task_poll
+            if duration <= 0:
+                self.fail_json(msg="Timeout waiting for Task {0}".format(task['id']))
+            time.sleep(self.task_poll)
+
+            resource_payload = self._resource_prepare_params('foreman_tasks', 'show', {'id': task['id']})
+            task = self._resource_call('foreman_tasks', 'show', resource_payload)
+        if not ignore_errors and task['result'] != 'success':
+            self.fail_json(msg='Task {0}({1}) did not succeed. Task information: {2}'.format(task['action'], task['id'], task['humanized']['errors']))
+        return task
+
+    def ping(self):
+        return self.resource('ping').call('ping')
+
+
+@pytest.fixture(scope="module")
+def foremanapi():
+    return ForemanApi(
+        uri=f'http://quadlet.tanso.example.com:{FOREMAN_PORT}',
```

Comment:
`pytest` runs on the "outer" system (my laptop), so it needs to be able to access Foreman like that.

probably clenest would be to have a *second* VM and run pytest (for the remote tests) there?

Reply:
Why is that?

Reply:
> `pytest` runs on the "outer" system (my laptop), so it needs to be able to access Foreman like that.

In smoker we use pytest-variables: https://github.com/theforeman/smoker?tab=readme-ov-file#variables. These are then used in https://github.com/theforeman/smoker/blob/d30ff41ca4931d7db61e8495c226f778b563a744/tests/conftest.py#L65-L73 to return an API instance.

> probably clenest would be to have a _second_ VM and run pytest (for the remote tests) there?

It's what we do in our pipelines with smoker.

Reply:
> Why is that?

Why is what? :)
Why I need to access from the "outside"?



> In smoker we use pytest-variables

Yeah, but someone/something needs to file in those variables.

I used paramiko to parse the ssh config we generate for testinfra for now.